File: go/AddTwoNumbers/AddTwoNumbers.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time -> 0(max(M, N)) Space -> 0(1)
func addTwoNumbers(l1, l2 *ListNode) *ListNode {
	carry := 0
	head := l1
	prev := l1
	for l1 != nil && l2 != nil {
		sum := l1.Val + l2.Val + carry
		val := sum % 10
		carry = sum / 10
		l1.Val = val
		prev = l1
		l1 = l1.Next
		l2 = l2.Next
	}

	if l1 != nil || l2 != nil {
		if l1 == nil {
			prev.Next = l2
			l1 = prev.Next
		}
		for carry != 0 && l1 != nil {
			sum := l1.Val + carry
			val := sum % 10
			carry = sum / 10
			l1.Val = val
			prev = l1
			l1 = l1.Next
		}
	}
	if carry != 0 {
		prev.Next = &ListNode{Val: carry}
	}

	return head
}

File: python/addTwoNumbers.py
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


# Time -> 0(max(M, N)) Space -> 0(1)
def addTwoNumbers(l1, l2) -> ListNode:
    head = l1
    prev = l1
    carry: int = 0
    while l1 is not None and l2 is not None:
        sum = l1.val + l2.val + carry
        val = sum % 10
        carry = sum // 10
        l1.val = val
        prev = l1
        l1 = l1.next
        l2 = l2.next

    if l1 is not None or l2 is not None:
        if l1 is None:
            prev.next = l2
            l1 = prev.next

        while carry != 0 and l1 is not None:
            sum = l1.val + carry
            val = sum % 10
            carry = sum // 10
            l1.val = val
            prev = l1
            l1 = l1.next

    if carry != 0:
        prev.next = ListNode(val=carry)
    return head
